src/main/java/com/fasterxml/jackson/databind/type/TypeFactory.java

```diff
@@ -480,23 +480,82 @@ private TypeBindings _bindingsForSubtype(JavaType baseType, int typeParamCount,
         // (hopefully passing null Class for root is ok)
         int baseCount = baseType.containedTypeCount();
         if (baseCount == typeParamCount) {
-            if (typeParamCount == 1) {
-                return TypeBindings.create(subclass, baseType.containedType(0));
-            }
-            if (typeParamCount == 2) {
-                return TypeBindings.create(subclass, baseType.containedType(0),
-                        baseType.containedType(1));
-            }
-            List<JavaType> types = new ArrayList<JavaType>(baseCount);
+            // 2017-10-17 epollan [databind#1604]
+            // Given "stacks" of type bindings contained in `baseType`; e.g. {@code baseType : Foo<A<B>, X>},
+            // with bindings A[B] (indicating a type binding to the JavaType for `A` with a nested
+            // type binding to `B`) and X; pop off all type bindings that might be implied in the
+            // `subclass`'s definition relative to the `baseType`.
+            // E.g. {@code subclass : Bar<B, X>} where {@code Bar<T, U> extends Foo<A<T>, U>}, would
+            // inherit type bindings B and X.  It would _not_ inherit `baseType`'s type binding to A[B]
+            // -- it would only inherit the inner type binding to B.
+            // This is because {@code Bar<?, ?>} already _is a_ {@code Foo<A<?>, ?>}, and serialization config
+            // for the class itself already comprehends the types that the inheritance specification
+            // has structurally locked into the base type's bindings.
+            JavaType[] bindings = new JavaType[baseCount];
             for (int i = 0; i < baseCount; ++i) {
-                types.add(baseType.containedType(i));
+                bindings[i] = _bindingForSubtypeAtBindingPosition(baseType, i, subclass);
             }
-            return TypeBindings.create(subclass, types);
+            return TypeBindings.create(subclass, bindings);
         }
         // Otherwise, two choices: match N first, or empty. Do latter, for now
         return TypeBindings.emptyBindings();
     }
 
+    private JavaType _bindingForSubtypeAtBindingPosition(JavaType baseType, int bindingPosition, Class<?> subclass) {
+        JavaType binding = baseType.containedType(bindingPosition);
+        // While the type binding is nested with exactly one inner binding, AND the subtype's definition
+        // itself already implies the outermost binding, then...
+        while (binding.containedTypeCount() == 1 && _subtypeImpliesBinding(baseType, binding, bindingPosition, subclass)) {
```

Comment:
This new code path is only activated when a top-level binding type is, itself, bound in a nested fashion to exactly one "inner" type

Reply:
It just so happens that this can be done for each type binding in sequence -- no combinatorics involved.